### PR TITLE
[tensorflow] [Deabstraction] Fix an assertion failure when it comes to non-constant attribute operands.

### PR DESCRIPTION
**Pull request by dingobye**
~~When collecting `constants` from operands of tensorOps, all TensowFlow Values are skipped. Such strategy is not conservative and may skip the operands which appear as TensorFlowValue but are indeed used as attributes. Consequently, assertion failures ([SR-8299](https://bugs.swift.org/browse/SR-8299))  may occur.~~

~~This patch adds considerations of `OperandClass`, making the constant collection more conservative.~~

When performing constant scanning for operands of tensorOps, all TensowFlow Values are not collected. In this case, we should not assert such operands always appear in `constants`. This patch fixes such assertion.

Resolves [SR-8299](https://bugs.swift.org/browse/SR-8299).

### CODE REVIEW

**Comment by dingobye** [comment]
@bgogul Hi Gogul, would you have a look at this, please?

**Comment by bgogul** [comment]
@swift-ci please test tensorflow

File: lib/SILOptimizer/Mandatory/TFDeabstraction.cpp [hunk 1]
```diff
@@ -1642,14 +1642,17 @@ void TFDeabstraction::checkAttributesAndFormGraphOps() {
   SmallVector<SILValue, 32> valuesToCheck;
 
   for (auto *op : tensorOps) {
-    for (auto &operand : op->getAllOperands()) {
+    auto opInfo = SILTensorOpInfo::decode(op);
```
Comment (bgogul) [inline lib/SILOptimizer/Mandatory/TFDeabstraction.cpp]: Should we assert that opInfo is valid?

**Review by bgogul — APPROVED** [review]
Nice! Thanks for the fix.

File: lib/SILOptimizer/Mandatory/TFDeabstraction.cpp (hunk 1 above)
Comment (dingobye) [reply lib/SILOptimizer/Mandatory/TFDeabstraction.cpp]: Thank you for the review, Gogul!

I tried asserting `opInfo` to be non-null here, however it breaks a number of tests... So I think we'd better go with the more robust way (i.e., testing its validity at [this line](https://github.com/apple/swift/pull/18249/files/054e230c322f2ef447b8db3161e49da0507da598#diff-d86f33ececb50054de2a79dcc8a95642R1652)).

Comment (bgogul) [reply lib/SILOptimizer/Mandatory/TFDeabstraction.cpp]: Thanks, Ding. In that case, let us leave it as is for this PR. 

However, isn't the failing tests an indication that we have collecting non tensor operations in tensorOps? Should we file a bug?

**Review by mhong — APPROVED** [review]
Thanks for the fix!

cc @lattner as FYI.

**Review by rxwei — APPROVED** [review]

**Comment by lattner** [comment]
Thank you for tackling this, I didn't consider the invalid case.  I'd prefer a different fix though - instead of adding the tensor operands to the things to evaluate, we should just fix the assertion point.  One moment:

File: lib/SILOptimizer/Mandatory/TFDeabstraction.cpp (hunk 1 above)
Comment (dingobye) [reply lib/SILOptimizer/Mandatory/TFDeabstraction.cpp]: Christ advised a different fix where `opInfo` is not involved. Anyway, I can take a look at this issue later.

[the remark this answers is not in the document]
Comment (dingobye) [reply lib/SILOptimizer/Mandatory/TFDeabstraction.cpp]: Will pay more attention to such style issues in future :-)

**Comment by dingobye** [comment]
Thank you for your feedback, Chris! I updated the code as you suggested.

**Review by lattner — APPROVED** [review]
Looks great!  Thank you again for the fix!

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Pull request merged**